### ISSUE TITLE
how to know whether a transaction committed or rollbacked inside a runtime hook 

### ISSUE THREAD

**Issue by guader** [ent/ent#2192]
In my case, I want to record the mutation to log or something just when the transaction is successfully committed.

Then I registered a runtime hook to the ent client like below:

```go
client.Use(func(next ent.Mutator) ent.Mutator {
        return ent.MutateFunc(func(ctx context.Context, m ent.Mutation) (ent.Value, error) {
		for _, f := range m.Fields() {
			v, ok := m.Field(f)
			fmt.Println(f, v, ok)
		}
		return next.Mutate(ctx, m)
	})
})
```

When I create a record in transaction for some type, say `Foo`, the hook was called.

Then I rollbacked the transaction with calling `tx.Rollback()` just after the creation, 

there was nothing in the `ctx` to indicate the result of the `tx`,

and the value of `ent.FooMutation.done` asserted from the `ent.Mutation` was `true`.

So I know the `done` field would not be set to `false` when a `tx` is rollbacked, but to indicate the `tx` is done regardless of it's result.

As I need to log the mutation with both `client.Foo.Create()` and `tx.Foo.Create()`, how can I do this without setting `tx.OnCommit()` everywhere?

**Comment by masseelch** [comment]
Hey @guader,

does the following fit your use case?

```go
client.Use(func(next ent.Mutator) ent.Mutator {
        return ent.MutateFunc(func(ctx context.Context, m ent.Mutation) (ent.Value, error) {
		for _, f := range m.Fields() {
			v, ok := m.Field(f)
		}
		if err := next.Mutate(ctx, m); err != nil {
      // error in mutation -> rolled back
   }
   // Committed
			fmt.Println(f, v, ok)
   return nil
	})
})
```

**Comment by a8m** [comment]
Hey @guader 👋

`ent.<T>Mutation.done` is an indicator for the mutation, not the transaction, as one transaction can execute multiple mutations.
You can use [transaction hooks](https://entgo.io/docs/transactions#hooks) with some additional helpers to log or store in the context every transaction information. 

```go
func OpenTx(ctx context.Context, c *ent.Client) (context.Context, *ent.Tx, error) {
    tx, err := client.Tx(ctx)
    if err != nil {
        return nil, nil err
    }
    // Add a hook on Tx.Commit.
    tx.OnCommit(func(next ent.Committer) ent.Committer {
        return ent.CommitFunc(func(ctx context.Context, tx *ent.Tx) error {
            // Code before the actual commit.
            err := next.Commit(ctx, tx)
            // Code after the transaction was committed.
            if err != nil {
                        // Log failure.
                        log.Println(...)
                        // Store info in context-key.
                        ctx.Value(ctxKey{}).(*TxInfo).Err = err
            }
            return err
        })
    })
    // Add a hook on Tx.Rollback.
    tx.OnRollback(func(next ent.Rollbacker) ent.Rollbacker {
        // ...
    })
    // Return a new context holds a *TxInfo to be updated by the
    // callbacks above.
    return context.WithValue(ctx, ctxKey{}, &TxInfo{}), tx, nil
}
```

This API can be generated as a client method or stored in the `./ent` directory as well.

**Comment by guader** [comment]
Thanks for your reply! @masseelch @a8m 

I've tested your @masseelch  resolution before with codes below:

```golang
func main() {
	ctx := context.Background()
	cli, _ := entity.Open("postgres", dataSrcName)
	defer func() { _ = cli.Close() }()
	_ = cli.Schema.Create(ctx)

	cli.Use(func(next ent.Mutator) ent.Mutator {
		return ent.MutateFunc(func(ctx context.Context, m ent.Mutation) (ent.Value, error) {
			val, err := next.Mutate(ctx, m)
			if err != nil {
				println(err.Error()) // issue NO ERR occurred here
				return val, err
			}
			for _, f := range m.Fields() {
				v, ok := m.Field(f)
				println(f, v.(string), ok) // print log here
			}
			return val, err
		})
	})

	tx, _ := cli.Tx(ctx)
	_ = tx.Foo.Create().SetName("foo").SaveX(ctx)
	_ = tx.Rollback()
}
```

and I've got the output as below:

```bash
~ $ go version
go version go1.17.1 darwin/amd64
~ $ grep ent go.mod
require entgo.io/ent v0.9.1
~ $ go run main.go
name foo true
````

so I think it's not the right way.

As for logging by `tx.OnCommit()` @a8m , a big problem is that we want to make the hook a common logging function for all the entities.

In fact, we are trying to trace the mutations applied to every record of every table in the database, it's something like an audit log with details.

Like, when someone updates `Foo.Name` from `foo` to `bar`, 
the hook prints some text log like `operator 'Alice' updated 'Foo.Name' from 'foo' to 'bar'`.

Maybe implementing an audit log on the orm is not a good idea though, and now we are aiming at implementing by triggers of postgresql.

So I think this issue could be closed, thanks, again!

**Comment by a8m** [comment]
> As for logging by tx.OnCommit() @a8m , a big problem is that we want to make the hook a common logging function for all the entities.
> 
> In fact, we are trying to trace the mutations applied to every record of every table in the database, it's something like an audit log with details.

Where are you logging these operations to? Same database or external service (or just stdout)? If you need to write them to the database, the operation will be transactional and you can verify this by using `m.Tx()` - we use the same approach. However, If you want to log committed mutations to an external service, I'd use the following approach:

1. Use the `OpenTx` function above, but add another field to the `TxInfo` - transaction logs buffer.
2. On each mutation (no matter before or after `Mutate()`), add a log to the `TxInfo` stored in the context.
3. `OnCommit` (using the `TxOpen` above) flush all logs on the buffer to the external service (or stdout). 

Does it make sense, or you need code snippets?

cc @hedwigz

**Comment by guader** [comment]
Thank you, It's very helpful of your advices @a8m

We are going to logging the operations just to the same database, but there are still some other issues to resolve.

(question1) We could use `OpenTx` to start a transaction, but what if we just use `client.Foo.Create()` to create an object? There are many codes written without using a transaction. shall we register another hook for `client.Foo.Create()` operation for logging purpose?

I've read comments of the interface `Mutation`, it says,

```golang
	// OldField returns the old value of the field from the database.
	// An error is returned if the mutation operation is not UpdateOne,
	// or the query to the database was failed.
	OldField(ctx context.Context, name string) (Value, error)
```

(question2) Does it mean I could get the value before updating only if I query the one updated first? 

Some codes like, `client.Foo.GetX(ctx, id).Update().SetName("bar").SaveX(ctx)`,

If I didn't query the very one first, I would not get any `Value` with `OldField` right?

(question3) What if I call `client.Foo.Update().Where(...).SaveX(ctx)` to do a batch updating? Does it call the hook for every object updated, or just once for the SQL statement? I need to logging the mutations for every object.

Thank you very much for your patience!

**Comment by a8m** [comment]
> There are many codes written without using a transaction. shall we register another hook for client.Foo.Create() operation for logging purpose?

Use one hook and control the logging based on tx info stored in the context. For example:

```go
client.Use(func(next ent.Mutator) ent.Mutator {
        return ent.MutateFunc(func(ctx context.Context, m ent.Mutation) (ent.Value, error) {
                v, err := next.Mutate(ctx, m)
                log(ctx, "<message>")
                return v, err
	})
})

// ...

func log(ctx context.Context, m string) {
    v := ctx.Value(ctxKey) // Or a custom call, e.g. `TxInfoFromContext(ctx)`.
    if v == nil {
        println(m)
    } else {
        v.(*TxInfo).Log(m)
    }
}
```

> Does it mean I could get the value before updating only if I query the one updated first?

Not, it fallbacks to the database in case the previous value does not exist in the mutation (e.g. `cient.T.UpdateOneID(id)...Save(ctx)`).

> (question3) What if I call client.Foo.Update().Where(...).SaveX(ctx) to do a batch updating? Does it call the hook for every object updated, or just once for the SQL statement? I need to logging the mutations for every object.

Use `Mutation.IDs` to get the affected IDs, because only one statement is executed on the database for updating the nodes.

> Thank you very much for your patience!

You're welcome! Closing, as the issue was resolved, but please, feel free to join our [Slack](https://entgo.io/docs/slack/) or [Discord](https://discord.gg/qZmPgTE6RX) communities 😎 ❤️

**Comment by guader** [comment]
Hi @a8m  ,

I've tested the `OpenTx()` with similar codes:

```golang
func BeginTx(ctx context.Context, c *entity.Client) (context.Context, *entity.Tx, error) {
	tx, err := c.Tx(ctx)
	if err != nil {
		return nil, nil, err
	}
	tx.OnCommit(func(next entity.Committer) entity.Committer {
		return entity.CommitFunc(func(ctx context.Context, tx *entity.Tx) error {
			if err := next.Commit(ctx, tx); err != nil {
				return err
			}
			fmt.Printf("CTX in OnCommit %+v\n", ctx) // print: CTX in OnCommit context.Background.WithValue(type string, val bar)
			ctx.Value(txInfoKey).(*TxInfo).Success = true
			return nil
		})
	})
	return context.WithValue(ctx, txInfoKey, &TxInfo{}), tx, nil
}

func main() {
	defer func() { _ = cli.Close() }()

	ctx := context.WithValue(context.Background(), "foo", "bar")
	ctx, tx, _ := BeginTx(ctx, cli)
	fmt.Printf("CTX in main %+v\n", ctx) // print: CTX in main context.Background.WithValue(type string, val bar).WithValue(type string, val <not Stringer>)
	_ = tx.Foo.Create().SetName("foo").SaveX(ctx)
	_ = tx.Commit()
}

// panic: interface conversion: interface {} is nil, not *main.TxInfo
```

And it seems that the `ctx` with `&TxInfo{}` were not passed to the `CommitFunc` I registered by `tx.OnCommit()`.

Take a look at these codes in the`main` func,

```golang
	ctx := context.WithValue(context.Background(), "foo", "bar")
	ctx, tx, _ := BeginTx(ctx, cli)
		...
	_ = tx.Foo.Create().SetName("foo").SaveX(ctx)
	_ = tx.Commit()
```

I think the reason should be the `tx.ctx` were not changed by `SaveX(ctx)`, when `tx.Commit()` was called, the `tx` still used the `ctx` set by `c.Tx(ctx)` in the function `BeginTx()`

This case shows that we couldn't get the context values passed to `SaveX()` right in the hook `tx.OnCommit()`, the context is not just passed through to the hook.

So there are still some improvements should be made, and I added `ctx = context.WithValue(ctx, txInfoKey, &TxInfo{})` to the very top of `BeginTx()`, then it works, they use the same context now.

BUT, please note that the runtime hook is called before the onCommit hook as `SaveX()` is called before `Commit()`.

So I can not get the value of `Success` from the context before the `tx` is commited, as the runtime hook is hard to decide whether to write a mutation log or not.

Did I still missed something for this approach?

I'll feedback if I get over it.

**Comment by guader** [comment]
ps:

> Use Mutation.IDs to get the affected IDs, because only one statement is executed on the database for updating the nodes.

There is no method `IDs` both for `Mutation` interface and `FooMutation` implementation, 

so with a batch update execution like `client.Foo.Update().Where(foo.NameEQ("foo")).SetName("bar").SaveX(ctx)`,

we could not know the effected row ids in a runtime hook,

and neither could I do a query with the predicates to get the ids as `predicates` is not exported from `FooMutaion`, 

besides, if we really do this with `reflect`, it's much effort to decide which query instance such as `FooQuery` to use in the runtime hook.

**Comment by guader** [comment]
FEEDBACK

We finally worked out of it, the points are:

1. set a sentry value in ctx to indicate a transaction
2. write log in runtime hook, but set the success status in runtime hook and onCommit hook, the sentry tell us which hook is responsible for the status setting
3. build `ent` cmd from master branch, and the `IDs` would be available in `v0.9.2`

**Comment by a8m** [comment]
Hey @guader, my code was incorrect in the comment above. The context should be created first, and then, the transaction should be created with it. See the updated version in https://github.com/ent/ent/issues/2192#issuecomment-984394417

Like you mentioned, updating ent to the latest master should solve the issue with IDs: `go get entgo.io/ent@master`. Please, let me know if I can help with anything else.
